Syscount libbpf

aspsk opened the pull request:
This PR converts BCC syscount to BPF CO-RE version and fixes some simple typos

aspsk commented:
Hi @anakryiko , here's a CO-RE port of `syscount`. One thing which is different from the original BCC-based `syscount` is that I am attaching both tracepoints even if there is no latency measurement requested. I am doing this because there is no official way to not load/attach a particular program using "skel" api. (It seems that it is probably hard to avoid loading the whole object, but it is pretty simple to skip attaching particular programs in the `bpf_object__attach_skeleton()` attach loop.)

anakryiko commented:
> Hi @anakryiko , here's a CO-RE port of `syscount`.

Thanks, will review shortly.

>One thing which is different from the original BCC-based `syscount` is that I am attaching both tracepoints even if there is no latency measurement requested. I am doing this because there is no official way to not load/attach a particular program using "skel" api. (It seems that it is probably hard to avoid loading the whole object, but it is pretty simple to skip attaching particular programs in the `bpf_object__attach_skeleton()` attach loop.)

Well, skeleton's attach() step is optional, you can attach each program individually with bpf_program__attach(skel->progs.whichever). 

It's true, though, that there's no way of skipping loading and verifying program right now. There was patch on the list to set specific BPF program to not auto-load, but it wasn't landed. I might add it shortly, it's really trivial and has come up before.

anakryiko commented on libbpf-tools/syscount.bpf.c:
```diff
@@ -0,0 +1,103 @@
+int tracepoint__raw_syscalls__sys_exit(struct trace_event_raw_sys_exit *args)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct data_t zero = {};
```
this is used for initialization only, right? So it's better to make it a static variable and skip zero-initializing it on every call

anakryiko commented on libbpf-tools/syscount.bpf.c:
```diff
@@ -0,0 +1,103 @@
+	if (args->id == -1)
+		return 0;
+
+	if (filter_pid && id >> 32 != filter_pid)
```
please parens around bit manipulation operations to avoid thinking about operator precedence

anakryiko commented on libbpf-tools/syscount.bpf.c:
```diff
@@ -0,0 +1,103 @@
+		return 0;
+
+	if (measure_latency) {
+		start_ts = bpf_map_lookup_elem(&start, &id);
```
there is no need to use entire 64-bits (tgid + tid) as a key, tid should be enough.

anakryiko commented on libbpf-tools/syscount.bpf.c:
```diff
@@ -0,0 +1,103 @@
+			return 0;
+	}
+
+	key = (count_by_process) ? id >> 32 : args->id;
```
you use (id >> 32) twice, I think it's better to have a local variable `tgid` instead.

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+	if (errno) {
+		warn("strtol: %s: %s\n", arg, strerror(errno));
+		return -1;
+	} else if (end == arg || val < min || val > max) {
```
wouldn't `*end == 0` be a better check that entire string argument was a valid integer, parsed completely?

aspsk replied:
Well, for example, here https://github.com/iovisor/bcc/pull/2920/files/3c40184699520362fece1d9ae3fe23d6182b9520#diff-68934a0eef57fae11fc290164fd99838R288 this is not the case. When I move the errno/syscall parsing to a separate file, this, probably, will. However, I also wanted to distinguish cases when errno is set vs. no symbols parsed. I will take a better look in the v2.

anakryiko replied:
link doesn't work, but sure, do whatever makes sense, I guess

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+}
+
+#ifdef __x86_64__
+static int errno_by_name_x86_64(const char *errno_name)
```
how about making it a bit cleaner and general by having a list of structs, like this:

```
static struct { int err_code; const char *err_name; } errs = {
    {1, "EPERM"},
    {2, "ENOENT"},
...
};
```
This seems like something that might be used by other tools eventually and will allow bi-directional lookups. code -> name and name -> code. And will avoid any unnecessary macro.

If necessary, it would also be possible to have (optionally) case-insensitive search.

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+	strcase("EOWNERDEAD", 130);
+	strcase("ENOTRECOVERABLE", 131);
+	strcase("ERFKILL", 132);
+	strcase("EHWPOISON", 133);
```
given it's so long, I'd extract it into a separate header file, honestly. Same for syscall names below. It distracts from the code flow enormously.

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+#endif
+
+/* Try to find the errno number using the errno(1) program */
+static int errno_by_name_dynamic(const char *errno_name)
```
is there a need for this? For unknown ENAME, why not just allowing specifying errno by integer value instead? This seems bloated and unnecessarily dangerous and complicated.

aspsk replied:
It is needed, unluckily, because the list I am compiling in is valid only for `x86_64`, not for other architectures. Funny that there's no reliable way to convert errno/syscalls <-> numbers, this should be provided by some library, probably.

anakryiko replied:
Sure, I don't object strongly, it just feels wrong to do popen from root-privileged tools for something so basic. Can't we generate similar lists for other architectures using the same approach?

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+	long x;
+
+	errno = 0;
+	x = strtol(buf, &end, 10);
```
use get_int?

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+ * (The idea is taken from src/python/bcc/syscall.py.)
+ */
+#ifdef __x86_64__
+static const char *syscall_names_x86_64[] = {
```
same as for errno, can you please split it into a separate file?

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+	int fd;
+	int n;
+
+	snprintf(path, sizeof(path), "/proc/%d/comm", pid);
```
this pid to comm mapping from procfs is racy, process could be dead or PID can even be reused in extreme cases. BPF code can easily capture this, so why not do that?

anakryiko replied:
ping? this approach doesn't work for short-lived processes at all. You can have a separate array for this or just extend start and store process name first time. Is there anything wrong with that?

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+	print_latency_header();
+	for (int i = 0; i < count && i < env.top; i++)
+		printf("%-22s %8llu %16.3lf\n",
+				agg_col(vals[i].key, buf, sizeof(buf)),
+				vals[i].count,
+				vals[i].total_ns / div);
```
formatting looks off, same in print_count. Is it just how Github renders it or it's really off?

aspsk replied:
Does this look better?
```
                col = agg_col(vals[i].key, buf, sizeof(buf));
                printf("%-22s %8llu %16.3lf\n", col, vals[i].count,
                                vals[i].total_ns / div);
```
and
```
                col = agg_col(vals[i].key, buf, sizeof(buf));
                printf("%-22s %8llu\n", col, vals[i].count);
```

anakryiko replied:
it's not about extracting agg_col call for me, it was about aligning each parameter with start of format string. It might be tabs vs spaces or you are using different indentation for 2nd+ arguments, don't know. But too lazy to pull your PR locally and check for myself :)

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,1167 @@
+		goto cleanup_obj;
+	}
+
+	err = syscount_bpf__attach(obj);
```
so here you don't have to attach everything, you can control which programs need to be attached and then use 
```
bpf_program__attach(obj->progs.whichone);
```

aspsk replied:
Right!

aspsk commented on libbpf-tools/syscount.bpf.c:
```diff
@@ -0,0 +1,103 @@
+	return 0;
+}
+
+SEC("tracepoint/raw_syscalls/sys_exit")
```
Am I right that in the case of raw tracepoints we are getting an array of `u64`s pointing to arguments of `TP_PROTO`? In this case prototype for `sys_exit` tracepoint is `TP_PROTO(struct pt_regs *regs, long ret)` and doesn't include the `id` (syscall number). In kernel this is easily obtained as `syscall_get_nr(current, regs)`, but I am not sure how to do this in the BPF case (as it is arch-dependent).

anakryiko replied:
Yeah, it's very strange, but sys_exit does not have syscall id provided... You are right that it seems to be arch-dependent. You can get syscall id using pt_regs->orig_ax, but that would break on non-x86. I guess just stick to tracepoint then.

aspsk commented:
Thanks for your comments @anakryiko . I will try to address all in the next version. (Except that I do not understand how to use a raw tracepoint in a portable way, as I described above.)

anakryiko commented:
> Thanks for your comments @anakryiko . I will try to address all in the next version. (Except that I do not understand how to use a raw tracepoint in a portable way, as I described above.)

Yeah, sys_exit not including syscall number is a downer, I wonder if it would make sense to fix it in kernel...

aspsk commented:
Hi @anakryiko , I've updated the PR, could you take another look?

One corresponding question: some `libbpf` functions, like `bpf_program__attach` return pointers, and, in case of error, they return kernel-style `ERR_PTR` pointers. However, the `linux/err.h` is not exported by `uapi`, so to use `libbpf` reliably one needs to redefine some of these macros, which is annoying to do for each new app. So, as we can't change API of the functions returning pointers, can't `libbpf` define these macros in `libbpf.h` instead of including `linux/err.h`?

anakryiko commented:
> Hi @anakryiko , I've updated the PR, could you take another look?
> 

sure

> One corresponding question: some `libbpf` functions, like `bpf_program__attach` return pointers, and, in case of error, they return kernel-style `ERR_PTR` pointers. However, the `linux/err.h` is not exported by `uapi`, so to use `libbpf` reliably one needs to redefine some of these macros, which is annoying to do for each new app. So, as we can't change API of the functions returning pointers, can't `libbpf` define these macros in `libbpf.h` instead of including `linux/err.h`?

libbpf defines libbpf_get_error() function, please use it.

aspsk commented:
> libbpf defines libbpf_get_error() function, please use it.

Ah, sorry, didn't see it. Updated the PR to use it.

anakryiko commented on libbpf-tools/syscall_helpers.c:
```diff
@@ -0,0 +1,526 @@
+close:
+	err = pclose(f);
+	if (err < 0)
+		warn("pcolse: %s\n", strerror(errno));
```
typo: pclose

aspsk replied:
<s>I fixed some typos in other tools in this PR, so I can add one!</s>

Ah, fixed. Thanks for reviewing @anakryiko!

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,466 @@
+ * measurements needed.
+ */
+static int attach_progs(struct syscount_bpf *obj)
+{
+	struct bpf_object_skeleton *s = obj->skeleton;
+	struct bpf_program *prog;
+	struct bpf_link **link;
+	const char *title;
+	int err;
+	int i;
+
+	for (i = 0; i < s->prog_cnt; i++) {
+		prog = *s->progs[i].prog;
+		link = s->progs[i].link;
+
+		title = bpf_program__title(prog, false);
+		if (!env.latency && strstr(title, "sys_enter"))
+			continue;
+
+		*link = bpf_program__attach(prog);
+		err = libbpf_get_error(*link);
+		if (err) {
+			warn("failed to auto-attach program '%s'\n",
+			     bpf_program__name(prog));
+			return err;
+		}
+	}
+
+	return 0;
+}
```
I don't get why you need this generic function. You have exactly 2 programs, and in some cases you want to attach only one of them. Here's the simplest and most straightforward way to do this:
```
obj->links.sys_exit = bpf_program__attach(obj->progs.sys_exit);
err = libbpf_get_error(obj->links.sys_exit);
if (err) /* ... */
if (env.latency) {
    obj->links.sys_enter = bpf_program__attach(obj->progs.sys_enter);
    err = libbpf_get_error(obj->links.sys_exit);
    if (err) /* ... */
}
```

Instead you are using obj->skeleton and use stuff that's meant for libbpf itself only. For no good reason, actually.

BTW, I'd call sys_enter/sys_exit programs something more short (I used just `sys_enter`/`sys_exit`) in the example above, so that it's easier to use from user-space.

aspsk replied:
> Instead you are using obj->skeleton and use stuff that's meant for libbpf itself only. For no good reason, actually.

You're right, I will fix this.

BTW, I've found a bug which I didn't see before (probably my system wasn't loaded enough, and also I've used batch version of lookup_and_delete initially before you rejected the same thing in vfsstat). There is a race in the `read_vals()`: the loop is not atomic and consists of multiple `bpf` system calls. Thus, the `bpf` entry in the map may be re-populated until it is the last entry in the map, and `bpf` entry could appear multiple times.

In the original tool, it does the printing as follows:
```
items = data.items() // read all values from the map one-by-one
keys = data.keys() // get all map keys
for k in keys: delete(k) // delete one element from the map
```
this also has some problems, because it looses system calls which are listed in keys and happened between getting items and deleting elements by keys.

The right solution which doesn't seem to loose any syscalls would be to do
```
keys = get_keys()
items = get_and_delete_items(keys)
```
but the `bpf_map_lookup_and_delete_elem` is supported only for stacks/queues.

Should I just use the approach used in syscount.py and loose some (relatively few) syscalls? Or maybe the right way is to patch `bpf_map_lookup_and_delete_elem` to support hashes? (Or there is some reason why it wasn't patched when batch lookup_and_delete was added?)

anakryiko replied:
Don't know about lookup_and_delete for hashmap, honestly.

I'd just use two hashmaps and switch between them for each reporting period. From user-space, before you start reporting, increment an index ("current map index"), re-routing BPF logic to a different map. Then work with the one that's not modified, read it, clear it, then on next reporting period bump counter again. On BPF side, before updating data map, you just:
```
int cur_map_idx = 0; /* this gets updated from user-space */

SEC(...)
int prog(...) {
    void *data = (cur_map_idx & 1) ? data1 : data2;
    /* proceed as before */
}
```

There is still tiny race (after you bumped cur_map_idx, there could be still BPF programs that are still updating previous map), but I think they are so tiny that it doesn't matter in practice.

BTW, to mitigate some of the memory overhead, I think it should be fine to use BPF_F_NO_PREALLOC hashmap (just specify it as map_flags). It will make kernel allocate memory for hash map entries on demand. So even with big hashmap, if it doesn't have lots of entries, you'd still have a manageable memory usage.

Makes sense?

aspsk replied:
So, if we still have a race, then I, probably, would prefer to read data the same way the original `syscount.py` — in this way I can re-use functions to read keys and hash items later. But I will try both variants first and see what looks better.

aspsk replied:
> BTW, to mitigate some of the memory overhead, I think it should be fine to use BPF_F_NO_PREALLOC hashmap (just specify it as map_flags). It will make kernel allocate memory for hash map entries on demand. So even with big hashmap, if it doesn't have lots of entries, you'd still have a manageable memory usage.

Thanks, will add this.

aspsk commented:
@anakryiko For some reason I can't answer in the thread: will fix the `"/proc/%d/comm"` issue, just forgot about it completely.

anakryiko commented on libbpf-tools/syscount.h:
```diff
@@ -0,0 +1,14 @@
+struct data_t {
+	__u64 count;
+	__u64 total_ns;
+	__u32 key;
```
btw, doesn't seem like you are using key field (and you don't need it, right?) 50% memory reduction right there ;)

aspsk replied:
I am using it in user space: I read items into an array of `struct data_t` and sort it by `count` or `total_ns`, but the `key` part should also be sorted to print it. This came from the "batch" version — this way I don't need any extra copy in user space. But now, comparing to syscall overhead this doesn't count, so I can remove this item and use a different data type in user space to store items.

aspsk commented:
Hi @anakryiko I think that I've addressed all your comments. I decided to read values as it is done in the original tool (read keys, read values by keys, delete values by keys), because it will be straightforward to fix this variant when `bpf_map_lookup_and_delete_batch` will be accessible and to correspond to the `syscount.py` beaviour. For command names I've added a LRU hash which I allocate from user space (`bpf_map__resize(obj->maps.comm, MAX_ENTRIES);`) only when needed. Also added some comments where I think it is not obvious why things are happening as they are.

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,455 @@
+
+static bool read_vals(int fd, struct data_ext_t *vals, __u32 *count)
+{
+	__u32 keys[MAX_ENTRIES] = { -1 };
```
that's quite a big on-the-stack array which is initialized every single time, make it static maybe?

aspsk replied:
Ok, thanks, there is actually no need to init it at all, removed.

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,455 @@
+	/* There is a race here: system calls which are represented by keys
+	 * above and happened between lookup and delete will be ignored.  This
+	 * will be fixed in future by using bpf_map_lookup_and_delete_batch,
+	 * but this function is too fresh to use it in bcc. */
```
this is not bcc, really :) you can use it, you'll be limiting yourself to very recent kernels, though

aspsk replied:
Ok, what if I do this based on `uname(2)` information (or there is a better way to get linux version)? I am curious if I can actually do race-less code using batch lookup and delete (or I need a `bpf_map_lookup_and_delete_elem` for hashes)

aspsk commented on libbpf-tools/syscount.bpf.c:
```diff
@@ -0,0 +1,133 @@
+	 * then first syscalls after fork, but before exec, run with the "bash"
+	 * comm, so if we cache this value here, then all processes created by
+	 * bash will be named "bash" */
+	err = bpf_get_current_comm(buf, sizeof(buf));
```
Yeah, thanks, I will read process name instead (which makes me think about adding a `-t` option to count/display per-thread statistics).

So, for rewriting, I need to do this, because a process/thread can change its name, say, using `prctl` or after `exec`. (Probably it could be done in sync, like, when we obtain a process name once and then watching for syscalls which can change it, but this sounds like a hard thing to implement on practice for a debug tool?)

aspsk commented on libbpf-tools/syscount.h:
```diff
@@ -0,0 +1,15 @@
+
+#define MAX_ENTRIES 8192
+
+#define COMM_NAME_SIZE 32
```
Yes, thanks. I was confused by `/proc/*/comm` which can return more bytes (for kernel tasks)

aspsk commented on libbpf-tools/syscount.bpf.c:
```diff
@@ -0,0 +1,133 @@
+	__uint(type, BPF_MAP_TYPE_LRU_HASH);
+	__type(key, u32);
+	__type(value, __u8[COMM_NAME_SIZE]);
+} comm SEC(".maps");
```
Ok, right

aspsk commented:
@anakryiko I've fixed all "blockers", didn't touch the race with getting values and the optimization with reading process names, planning to to this later (hopefully, together with adding per-thread view)

aspsk commented:
> @anakryiko I've fixed all "blockers", didn't touch the race with getting values and the optimization with reading process names, planning to to this later (hopefully, together with adding per-thread view)

Upd.: implemented reading values in a raceless way using batch functions, if available

anakryiko commented on libbpf-tools/syscount.c:
```diff
@@ -0,0 +1,482 @@
+		vals[j].count = val.count;
+		vals[j].total_ns = val.total_ns;
+		vals[j].key = keys[j];
+		strncpy(vals[j].comm, val.comm, TASK_COMM_LEN);
```
newer GCCs will complain about potentially non-zero-terminated strings. I'd just do straight memcpy of TASK_COMM_LEN bytes (which we know are zero-terminated).

aspsk replied:
ok, done

aspsk commented:
> LGTM. Let's land this beast.
> 
> This will probably conflict with cpudist, so one of you guys will need to rebase one more time, sorry about that!

Np, I will rebase mine once `cpudist` is merged (we will have a conflict in `Makefile`).

aspsk commented:
> > LGTM. Let's land this beast.
> > This will probably conflict with cpudist, so one of you guys will need to rebase one more time, sorry about that!
> 
> Np, I will rebase mine once `cpudist` is merged (we will have a conflict in `Makefile`).

Taking my words back @anakryiko! The `cpudist` already has conflicts, so this would be wise to merge my PR first :-)

yonghong-song commented:
[buildbot, test this please]
